### PR TITLE
1. Alt parser

### PR DESCRIPTION
**Pull request by huy-dna**
## Summary
* Add an alternative hand-written dbml parser (not integrated yet).
* How the parser's currently work: https://dbml-explorer-demo-1.vercel.app/
* Next PR: #419 

## Issue
[Asana](https://app.asana.com/0/1125458388096768/1205090003669536/f)

## Lasting Changes (Technical)

* Add an alternative dbml parser.

## Checklist

Please check directly on the box once each of these are done

- [ ] Documentation (if necessary)
- [ ] Tests (integration test/unit test)
- [ ] Integration Tests Passed
- [ ] Code Review

### CODE REVIEW

**Comment by NQPhuc** [comment]
Please ensure all CI tests passed

File: packages/dbml-core/src/parse/dbml/src/lib/lexer/lexer.ts [hunk 1]
```diff
@@ -0,0 +1,567 @@
...
+          this.addToken(SyntaxTokenKind.SEMICOLON);
+          break;
+        case ':':
+          this.advance();
+          this.addToken(SyntaxTokenKind.COLON);
+          break;
+        case '<':
+          this.advance();
+          if (this.peek() !== '=') {
+            this.addToken(SyntaxTokenKind.LT);
```
Comment (phuongduyphan) [inline packages/dbml-core/src/parse/dbml/src/lib/lexer/lexer.ts:153]: Instead of classifying for each operator `<=`, `>`, etc., should we use one kind `OPERATOR` for all of the operators? Doing that in the future if we want to support new operators, we don't need to parse again. Also, same operator in different contexts has different names/meaning (i.e. `>` this could be greater than in binary expression, but it could be many-to-one operator in relationship expression)

File: packages/dbml-core/src/parse/dbml/src/lib/lexer/lexer.ts [hunk 2]
```diff
@@ -0,0 +1,567 @@
...
+      case 'f':
+        char = 'f';
+        this.advance();
+        break;
+      case 'u': {
+        let hex = '';
+        for (let i = 0; i <= 3; i += 1) {
+          if (this.peek() && isAlphaNumeric(this.peek()!)) {
+            hex += this.advance();
+            char = String.fromCharCode(parseInt(hex, 16));
```
Comment (phuongduyphan) [inline packages/dbml-core/src/parse/dbml/src/lib/lexer/lexer.ts:552]: Should we put this outside of the loop?

File: packages/dbml-core/src/parse/dbml/src/lib/lexer/lexer.ts [hunk 3]
```diff
@@ -0,0 +1,567 @@
...
+          this.start - 3,
+          this.current - this.start + 6,
+          string,
+        ),
+      );
+      this.skip(3);
+    }
+  }
+
+  functionExpression() {
```
Comment (phuongduyphan) [inline packages/dbml-core/src/parse/dbml/src/lib/lexer/lexer.ts:335]: This method logic is quite similar to `quotedVariable`, `singleLineStringLiteral`, etc... Should we have a generic function for all of them?

File: packages/dbml-core/src/parse/dbml/src/lib/utils.ts [hunk 4]
```diff
@@ -0,0 +1,15 @@
+export function isAlpha(char: string): boolean {
+  const [c] = char;
+
+  return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || c === '_';
```
Comment (phuongduyphan) [inline packages/dbml-core/src/parse/dbml/src/lib/utils.ts:4]: why is `_` included in this check?

**Comment by phuongduyphan** [comment]
This PR is quite large @huy-dna, next time suggest we create multiple smaller PRs (i.e one for lexer, one for parser)

**Comment by phuongduyphan** [comment]
Also pls help me benchmark the new parser with the old pegjs one

File: packages/dbml-core/src/parse/dbml/src/lib/lexer/lexer.ts [hunk 5]
```diff
@@ -0,0 +1,567 @@
...
+        this.advance();
+        c = this.peek();
+      }
+      this.errors.push(
+        new ParsingError(ParsingErrorCode.INVALID, 'Invalid number', this.start, this.current),
+      );
+    } else {
+      this.addToken(SyntaxTokenKind.NUMERIC_LITERAL);
+    }
+  }
```
Comment (phuongduyphan) [inline packages/dbml-core/src/parse/dbml/src/lib/lexer/lexer.ts:445]: Why is this method so complex than it is in AML @huy-dna?

File: packages/dbml-core/src/parse/dbml/src/lib/lexer/lexer.ts [hunk 6]
```diff
@@ -0,0 +1,567 @@
...
+                ParsingErrorCode.EXPECTED_THINGS,
+                `Unexpected token ${c}`,
+                this.start,
+                this.current - 1,
+              ),
+            );
+            break;
+          }
+      }
+      this.closeGap();
```
Comment (phuongduyphan) [inline packages/dbml-core/src/parse/dbml/src/lib/lexer/lexer.ts:222]: What is the meaning of this?

File: packages/dbml-core/src/parse/dbml/src/lib/lexer/lexer.ts [hunk 7]
```diff
@@ -0,0 +1,567 @@
...
+        triviaList.push(token);
+        if (token.kind === SyntaxTokenKind.NEWLINE) {
+          if (lastNonTrivia) {
+            lastNonTrivia.trailingTrivia = triviaList;
+            startOfLine = true;
+            lastNonTrivia = undefined;
+            triviaList = [];
+          }
+        }
+        continue;
```
Comment (phuongduyphan) [inline packages/dbml-core/src/parse/dbml/src/lib/lexer/lexer.ts:247]: this code block has 3 nested if with a continue statement. I think the logic is quite hard to follow and understand what's going on here. It's great if you could add some notes for this, or refactor to a simpler logic if possible

File: packages/dbml-core/src/parse/dbml/src/lib/result.ts [hunk 8]
```diff
@@ -0,0 +1,34 @@
+import { ParsingError } from './errors';
+
+export default class Result<T> {
```
Comment (phuongduyphan) [inline packages/dbml-core/src/parse/dbml/src/lib/result.ts:3]: It's better to add some note to explain what are this `Result` class used for? Is it the result of the parsing process or it could be used for other generic use cases?

Comment (phuongduyphan) [reply packages/dbml-core/src/parse/dbml/src/lib/result.ts:3]: If it could be used for other generic use cases, does it make sense to hardcode the error type for this?

File: packages/dbml-core/src/parse/dbml/src/lib/result.ts [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+  private value?: T;
+
+  private errors: ParsingError[];
```
Comment (phuongduyphan) [inline packages/dbml-core/src/parse/dbml/src/lib/result.ts:6]: If it could be used for other generic use cases, does it make sense to hardcode the error type for this?

File: packages/dbml-core/src/parse/dbml/src/lib/errors.ts [hunk 10]
```diff
@@ -0,0 +1,28 @@
+export enum ParsingErrorCode {
```
Comment (phuongduyphan) [inline packages/dbml-core/src/parse/dbml/src/lib/errors.ts:1]: Why do we need this?

File: packages/dbml-core/src/parse/dbml/src/lib/lexer/lexer.ts (hunk 5 above)
Comment (huy-dna) [reply packages/dbml-core/src/parse/dbml/src/lib/lexer/lexer.ts:445]: This one also checks for adjacent alphabetical characters after the digits. If there are such characters, the number is marked invalid.
In AML, the lexer doesn't check this, which leads to the problem mentioned in this:
https://app.asana.com/0/1152839153201473/1205071037227754/f

File: packages/dbml-core/src/parse/dbml/src/lib/lexer/lexer.ts [hunk 11]
```diff
@@ -0,0 +1,567 @@
...
+    );
+  }
+
+  lex(): Result<SyntaxToken[]> {
+    this.init();
+    while (!this.isAtEnd()) {
+      const c: string = this.peek()!;
+      switch (c) {
+        case ' ':
+          this.advance();
```
[the remark this answers is not in the document]
Comment (huy-dna) [reply packages/dbml-core/src/parse/dbml/src/lib/lexer/lexer.ts:68]: I agree with the `skip()` method. About the repetition of `advance()`, I have thought about it before. However, at the same time, I would like methods such as `identifier`, `functionExpression` to be standalone, independent of the context they were called -- not that it was called in a situation where the first character has been consumed by `advance()`.

File: packages/dbml-core/src/parse/dbml/src/lib/errors.ts (hunk 10 above)
Comment (huy-dna) [reply packages/dbml-core/src/parse/dbml/src/lib/errors.ts:1]: It's a way to map a type of error to a code. It isn't strictly needed and I only include it if this may be desired in the future.

File: packages/dbml-core/src/parse/dbml/src/lib/lexer/lexer.ts (hunk 3 above)
Comment (huy-dna) [reply packages/dbml-core/src/parse/dbml/src/lib/lexer/lexer.ts]: I have added a method called extractString

**Comment by huy-dna** [comment]
> Also pls help me benchmark the new parser with the old pegjs one

I have performed a cursory benchmark for the two:
For the following program:
image

Running it for about ten times, the total time of:
* The pegjs one:
  
  image

* The new one:
  
  image

**Comment by huy-dna** [comment]
Benchmark results:

* 18k
  * Pegjs (5 times):
  
    image

  * New one (10 times):
  
    image

* 25k
  * Pegjs (5 times):
     
   image


  * New one (10 times):
    
    image

File: packages/dbml-core/src/parse/dbml/src/lib/lexer/lexer.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (phuongduyphan) [reply packages/dbml-core/src/parse/dbml/src/lib/lexer/lexer.ts:71]: > I would like methods such as identifier, functionExpression to be standalone, independent of the context they were called

So why do you prefer this em? Is there any specific reason behind?

File: packages/dbml-core/src/parse/dbml/src/lib/utils.ts (hunk 4 above)
Comment (phuongduyphan) [reply packages/dbml-core/src/parse/dbml/src/lib/utils.ts:4]: If so I don't think the name of this function reflects that e

File: packages/dbml-core/src/parse/dbml/src/lib/lexer/lexer.ts (hunk 5 above)
Comment (phuongduyphan) [reply packages/dbml-core/src/parse/dbml/src/lib/lexer/lexer.ts:379]: Oh I see. But I don't think the lexer should care about it. IMO, this logic should be in the parser. TypeScript is still able to scan this one `2const` as two separate tokens `2` and `const`

File: packages/dbml-core/src/parse/dbml/src/lib/errors.ts (hunk 10 above)
Comment (phuongduyphan) [reply packages/dbml-core/src/parse/dbml/src/lib/errors.ts:1]: What are the use cases in the future that would need this em?

File: packages/dbml-core/src/parse/dbml/src/lib/parser/parser.ts [hunk 12]
```diff
@@ -0,0 +1,827 @@
...
+    this.contextStack = new ParsingContextStack();
+  }
+
+  private advance(): SyntaxToken {
+    return this.tokens[this.current++];
+  }
+
+  private peek(lookahead: number = 0): SyntaxToken | undefined {
+    if (lookahead + this.current >= this.tokens.length) {
+      return undefined;
```
Comment (phuongduyphan) [inline packages/dbml-core/src/parse/dbml/src/lib/parser/parser.ts:64]: Instead of returning undefined, could we consider returning the last EOF token instead? By doing this, the return type of our method becomes uniform, and we can eliminate the need for an undefined check every time we invoke it.

File: packages/dbml-core/src/parse/dbml/src/lib/parser/parser.ts [hunk 13]
```diff
@@ -0,0 +1,827 @@
...
+
+  private init() {
+    this.current = 0;
+    this.errors = [];
+    this.invalid = [];
+    this.contextStack = new ParsingContextStack();
+  }
+
+  private advance(): SyntaxToken {
+    return this.tokens[this.current++];
```
Comment (phuongduyphan) [inline packages/dbml-core/src/parse/dbml/src/lib/parser/parser.ts:59]: Should we handle the case when `this.current` is out of bound?

File: packages/dbml-core/src/parse/dbml/src/lib/parser/parser.ts [hunk 14]
```diff
@@ -0,0 +1,827 @@
...
+} from './nodes';
+
+export default class Parser {
+  private tokens: SyntaxToken[];
+
+  private current: number = 0;
+
+  private errors: ParsingError[] = [];
+
+  private invalid: (SyntaxToken | SyntaxNode)[] = [];
```
Comment (phuongduyphan) [inline packages/dbml-core/src/parse/dbml/src/lib/parser/parser.ts:37]: What is this `invalid` used for?

File: packages/dbml-core/src/parse/dbml/src/lib/parser/parser.ts [hunk 15]
```diff
@@ -0,0 +1,827 @@
...
+      throw error;
+    }
+  }
+
+  parse(): Result<SyntaxNode & { kind: SyntaxNodeKind.PROGRAM }, ParsingError> {
+    const body: ElementDeclarationNode[] = [];
+
+    this.init();
+
+    while (this.peek() && this.peek()?.kind !== SyntaxTokenKind.EOF) {
```
Comment (phuongduyphan) [inline packages/dbml-core/src/parse/dbml/src/lib/parser/parser.ts:113]: Can we refactor this line to `while (!this.isAtEnd) {`?

File: packages/dbml-core/src/parse/dbml/src/lib/parser/parser.ts [hunk 16]
```diff
@@ -0,0 +1,827 @@
...
+      try {
+        body.push(this.elementDeclaration());
+      } catch (e) {
+        if (!(e instanceof ParsingError)) {
+          throw e;
+        }
+        if (!(this.peek()?.kind === SyntaxTokenKind.EOF)) {
```
Comment (phuongduyphan) [inline packages/dbml-core/src/parse/dbml/src/lib/parser/parser.ts:120]: same here, can we use `!this.isAtEnd()`?

File: packages/dbml-core/src/parse/dbml/src/lib/parser/nodes.ts [hunk 17]
```diff
@@ -0,0 +1,575 @@
+import { SyntaxToken } from '../lexer/tokens';
+
+export interface SyntaxNode {
+  kind: SyntaxNodeKind;
+  start: number;
```
Comment (phuongduyphan) [inline packages/dbml-core/src/parse/dbml/src/lib/parser/nodes.ts:5]: Does this start count from node's leading trivia (comments, etc.) or from the node's start itself?

File: packages/dbml-core/src/parse/dbml/src/lib/parser/nodes.ts [hunk 18]
```diff
@@ -0,0 +1,575 @@
...
+  | ListExpressionNode
+  | TupleExpressionNode
+  | CallExpressionNode
+  | PrimaryExpressionNode
+  | FunctionExpressionNode
+  | AccessExpressionNode;
+
+export type ExpressionNode =
+  | LiteralElementExpressionNode
+  | NormalFormExpressionNode
```
Comment (phuongduyphan) [inline packages/dbml-core/src/parse/dbml/src/lib/parser/nodes.ts:212]: This `NormalFormExpressionNode` also contains `LiteralElementExpressionNode` and `FunctionApplicationNode`, so what is the point of creating a new type `ExpressionNode`? :pepe-facepalm

File: packages/dbml-core/src/parse/dbml/src/lib/parser/nodes.ts [hunk 19]
```diff
@@ -0,0 +1,575 @@
...
+        this.end = lastName.offset + lastName.length - 1;
+      }
+    }
+    this.name = name;
+    this.valueOpenColon = valueOpenColon;
+    this.value = value;
+  }
+}
+
+export type NormalFormExpressionNode =
```
Comment (phuongduyphan) [inline packages/dbml-core/src/parse/dbml/src/lib/parser/nodes.ts:196]: What does "normal form expression" mean?

File: packages/dbml-core/src/parse/dbml/src/lib/parser/parser.ts [hunk 20]
```diff
@@ -0,0 +1,829 @@
...
+      this.peek()?.kind !== SyntaxTokenKind.LBRACE
+    ) {
+      try {
+        name = this.normalFormExpression();
+      } catch (e) {
+        if (e instanceof ParsingError) {
+          this.synchronizeElementDeclarationName();
+        } else {
+          throw e;
+        }
```
Comment (phuongduyphan) [inline packages/dbml-core/src/parse/dbml/src/lib/parser/parser.ts:156]: This logic is duplicated in multiple places, suggest we should have a dedicated function for this

File: packages/dbml-core/src/parse/dbml/src/lib/parser/parser.ts [hunk 21]
```diff
@@ -0,0 +1,829 @@
...
+
+    let attributeList: ListExpressionNode | undefined;
+    if (this.check(SyntaxTokenKind.LBRACKET)) {
+      attributeList = this.listExpression();
+    }
+
+    let body: ExpressionNode | BlockExpressionNode | undefined;
+    let bodyOpenColon: SyntaxToken | undefined;
+
+    if (!this.check(SyntaxTokenKind.COLON, SyntaxTokenKind.LBRACE)) {
```
Comment (phuongduyphan) [inline packages/dbml-core/src/parse/dbml/src/lib/parser/parser.ts:182]: how about checking in this order em: 
```
if (this.match(SyntaxTokenKind.COLON)) {
  ...
} else if (this.match(SyntaxTokenKind.LBRACE)) {
  ...
} else {
  ...
}
```

IMO, checking in this order is more natural and easier to follow than yours

File: packages/dbml-core/src/parse/dbml/src/lib/parser/nodes.ts [hunk 22]
```diff
@@ -0,0 +1,575 @@
...
+    body: BlockExpressionNode | ExpressionNode;
+  }) {
+    this.start = type.offset;
+    this.end = body.end;
+    this.type = type;
+    this.name = name;
+    this.as = as;
+    this.alias = alias;
+    this.attributeList = attributeList;
+    this.bodyOpenColon = bodyOpenColon;
```
Comment (phuongduyphan) [inline packages/dbml-core/src/parse/dbml/src/lib/parser/nodes.ts:112]: where does `bodyOpenBrace` and `bodyCloseBrace` get set em?

File: packages/dbml-core/src/parse/dbml/src/lib/errors.ts (hunk 10 above)
Comment (huy-dna) [reply packages/dbml-core/src/parse/dbml/src/lib/errors.ts:1]: This is not very useful at current as it's still rudimentary. You can think of it's like HTTP 404 error code. AML has error codes, too :v.

File: packages/dbml-core/src/parse/dbml/src/lib/parser/nodes.ts (hunk 17 above)
Comment (huy-dna) [reply packages/dbml-core/src/parse/dbml/src/lib/parser/nodes.ts:5]: It counts from the content start, not including the trivia. It defines the start of the yellow highlight in the playground. Should I name it more explicitly?

File: packages/dbml-core/src/parse/dbml/src/lib/utils.ts [hunk 23]
```diff
@@ -0,0 +1,92 @@
...
+  expression: VariableNode & { variable: { kind: SyntaxTokenKind.IDENTIFIER } };
+} {
+  return (
+    value instanceof PrimaryExpressionNode &&
+    value.expression instanceof VariableNode &&
+    value.expression.variable.kind === SyntaxTokenKind.IDENTIFIER
+  );
+}
+
+export function last<T>(array: T[]): T | undefined {
```
Comment (phuongduyphan) [inline packages/dbml-core/src/parse/dbml/src/lib/utils.ts:86]: One tip: for those utility list functions we could use [lodash](https://lodash.com/) instead of rewriting e
